CorsixTH/Lua/app.lua

```diff
@@ -1680,57 +1680,131 @@ function App:loadLuaFolder(dir, no_results, append_to)
+-- attached e.g. 0.66.0-dev172
+function App:getVersion(savegame_version, concatenate)
+  savegame_version = savegame_version or self:getCurrentVersion()
+  local result
+  for i=#version_table, 1, -1 do
+    local version = version_table[i]
+    if version.version == savegame_version then
+      result = version
+      break
+    elseif version.version - savegame_version < 0 then
+      -- we're not on a release version
+      local devel_version = shallow_clone(version) -- prevent recursion
+      devel_version.patch = version.patch .. "-dev" .. savegame_version
+      result = devel_version
+      break
+    end
+  end
+  if concatenate and result then
+    if string.len(result.patch) > 0 then
+      result = "v" .. result.major .. "." .. result.minor .. "." ..
+          result.revision .. result.patch
+    else
+      result = "v" .. result.major .. "." .. result.minor .. "." .. result.revision
+    end
+  end
+  return result
+end
+
+--! Reports a difference between two versions based on requested methodology.
+--!param version_a (number or table) The first (usually newer) version to test
+--!param version_b (number or table) The second (usually older) version to test
+--!param method (string) What method to compare by
+--- method(release) reports the difference between the matching releases in steps, revisions are not counted.
+--- method(version) reports the difference between two savegame versions
+--- method(age) reports the difference in days between the matching release dates
+--! if a match is detected as a development version it will target the release it was built upon, except for via version method.
+function App:compareVersions(version_a, version_b, method)
+  assert(type(version_a) ~= "string" and type(version_b) ~= "string",
+      "Compare requires savegame version or the version table to function")
+  assert(method == "age" or method == "release" or method == "version",
+      "Not using a valid compare method")
+
+  if type(version_a) == "number" then
+    version_a = self:getVersion(version_a)
+  end
+  if type(version_b) == "number" then
+    version_b = self:getVersion(version_b)
+  end
+
+  if method == "release" then
+    local function countBackward(version_to_check)
+      local step = 0
+      for i=#version_table, 1, -1 do
+        local version = version_table[i]
+        if version.version == version_to_check then
+          break
+        elseif (version.version - version_to_check) < 0 then
+          -- we're not on a release version
+          if step == 0 then break end -- working from current development
+          step = step - 1
+          break
+        end
+        step = (version.revision == 0 and string.len(version.patch) == 0)
+            and step - 1 or step
+      end
+      return step
+    end
+    return countBackward(version_a.version) - countBackward(version_b.version)
+  end
+
+  if method == "version" then
+    return version_a.version - version_b.version
+  end
+
+  if method == "age" then
```

Comment:
Like version, this compare will only look at the table entries, it doesn't consider a development build any different in release date to the actual base release version. This could probably be scrapped, along with the date entries in the table (unless they're nice to have for some other purpose).

Reply:
It's somewhat a waste to delete information, even if not used I'd keep it.
Moving it to comment is an option, although it's harmless to have it in the table too.